system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+        FAR char *str = va_arg(ap, FAR char *);
+        ret = set_string(setting, str);
+      }
+      break;
+
+    case SETTING_INT:
+      {
+        i = va_arg(ap, int);
+        ret = set_int(setting, i);
+      }
+      break;
+
+    case SETTING_BYTE:
+      {
+        i = va_arg(ap, int);
+        ret = set_byte(setting, i);
+      }
+      break;
+
+    case SETTING_BOOL:
+      {
+        i = va_arg(ap, int);
+        ret = set_bool(setting, i);
+      }
+      break;
+
+    case SETTING_FLOAT:
+      {
+        double f = va_arg(ap, double);
+        ret = set_float(setting, f);
+      }
+      break;
+
+    case SETTING_IP_ADDR:
+      {
+        FAR struct in_addr *ip = va_arg(ap, FAR struct in_addr *);
+        ret = set_ip(setting, ip);
+      }
+      break;
+
+    default:
+      {
+        DEBUGASSERT(0);
+      }
+      break;
+  }
+
+  va_end(ap);
+
+  if (ret >= 0)
+    {
+      h = hash_calc();
+      if (h != g_settings.hash)
+        {
+          g_settings.hash = h;
+
+          signotify();
+          save();
+        }
+
+#if 0
+      /* It might help to know the value didn't actually change? */
+
+      else
+        {
+          ret = -EEXIST;
+        }
+#endif
+    }
+
+errout:
+  pthread_mutex_unlock(&g_settings.mtx);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: settings_iterate
+ *
+ * Description:
+ *    Gets a copy of a setting at the specified position. It can be used to
+ *    iterate over the settings map, by using successive values of idx.
+ *
+ * Input Parameters:
+ *    idx         - the iteration index for the setting.
+ *    setting     - pointer to return the setting value
+ *
+ * Returned Value:
+ *    Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_iterate(int idx, FAR setting_t *setting)
+{
+  int ret;
+
+  DEBUGASSERT(setting);
+
+  if ((idx < 0) || (idx >= CONFIG_SYSTEM_SETTINGS_MAP_SIZE))
+    {
+      memset(setting, 0, sizeof(setting_t));
+      return -EINVAL;
+    }
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (!g_settings.initialized)
+    {
+      ret = -EACCES;
+      goto errout;
+    }
+
+  memcpy(setting, &map[idx], sizeof(setting_t));
+
+errout:
+  pthread_mutex_unlock(&g_settings.mtx);
+
+  if (map[idx].type == SETTING_EMPTY)
+    {
+      return -ENOENT;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: settings_savepending
+ *
+ * Description:
+ *    Check if a save is pending. Useful if cached saves are in play.
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *    whether a save is pending: true or false;
+ *
+ ****************************************************************************/
+
+bool settings_savepending(void)
+{
+  bool wrpend;
+
+  wrpend = pthread_mutex_lock(&g_settings.mtx);
```

Comment:
It is possible that this function (and the use of the mutex that it does here) is the root of the problems you had with the recursive mutex. I need to think about it...

Reply:
It was simply wrong, and should not have been like that, ever. But it is no longer there as the savepending function is no more.